src/hotspot/share/opto/vector.cpp

```diff
@@ -252,6 +261,22 @@ void PhaseVector::scalarize_vbox_node(VectorBoxNode* vec_box) {
 #endif // ASSERT
                                                first_ind, n_fields);
     sobj->init_req(0, C->root());
+
+    // If a mask is feeding into a safepoint, then its value should be
+    // packed into a byte vector first, this will simplify the re-materialization
+    // logic for both predicated and non-predicated targets.
+    bool is_mask = is_vector_mask(iklass);
+    if (is_mask) {
+      if (vec_value->Opcode() == Op_VectorLoadMask) {
+        const Type* in1_ty = vec_value->in(1)->bottom_type();
+        assert(in1_ty->isa_vect() && in1_ty->is_vect()->element_basic_type() == T_BOOLEAN, "");
+        vec_value = vec_value->in(1);
+      } else if (vec_value->Opcode() != Op_VectorStoreMask) {
+        const TypeVect* vt = vec_value->bottom_type()->is_vect();
+        BasicType bt = vt->element_basic_type();
+        vec_value = gvn.transform(VectorStoreMaskNode::make(gvn, vec_value, bt, vt->length()));
```

Comment:
`VectorStoreMaskNode::Identity()` already handles `VectorStoreMask (VectorLoadMask bv) elem_size ==> bv` transformation, doesn't it? So, you can just unconditionally instantiate `VectorStoreMaskNode` and let IGVN handle it.

Also, an observation on naming: `VectorLoadMask` and `VectorStoreMask` names are misleading. On the surface, they look like memory nodes, but in fact, are cast nodes (`Vector <-> Mask`). Please, file an RFE to address that eventually.